### PR TITLE
Drop-down info tables for datasets and models; feature descriptions

### LINKED ISSUE #121: Add descriptive drop downs for datasets and models
Like the ones for features and predictions.
- For datasets, show the list of files (+ any metadata)?
- For models, show type, hyperparameters, etc. (accuracy?)

### PR DESCRIPTION
**Pull request by acrellin**
Closes #121 & #125

### CODE REVIEW

**Comment by acrellin** [comment]
Added displaying feature descriptions to this PR

**Comment by acrellin** [comment]
Ready for review (and should start passing when https://github.com/cesium-ml/cesium/pull/221 is merged)

**Comment by acrellin** [comment]
@stefanv ready for review whenever you have some time.

File: cesium_app/handlers/feature_list.py [hunk 1]
```diff
@@ -1,10 +1,23 @@
...
-
-from cesium.features import feature_categories, feature_tags
+from cesium.features.graphs import (feature_categories, feature_tags,
+                                    dask_feature_graph, extra_feature_docs)
 
 
 class FeatureListHandler(BaseHandler):
     def get(self):
+        feature_descriptions = {}
+        for feature_name in [e for e in dask_feature_graph
```
Comment (bnaul) [inline cesium_app/handlers/feature_list.py]: Looks good except for this bit which could be cleaner. How about:

```
def _get_docstring(func):
    return ...

feature_descriptions = {f: extra_feature_docs[f] if f in extra_feature_docs
                        else _get_docstring(dask_feature_graph[f][0]}
```

File: cesium_app/handlers/model.py [hunk 2]
```diff
@@ -102,14 +105,17 @@ def post(self):
             featureset=fset_data, model_type=model_type,
             model_parameters=model_params,
             params_to_optimize=params_to_optimize)
-        future = executor.submit(joblib.dump, computed_model, model_file.uri)
-        closed = executor.submit(xr.Dataset.close, fset_data)
+        score_future = executor.submit(build_model.score_model, computed_model,
```
Comment (bnaul) [inline cesium_app/handlers/model.py:108]: @stefanv these changes were our attempt at getting ensure that all the steps are always executed without any race conditions; can you think of a better way?

**Comment by bnaul** [comment]
👍 @stefanv ?

**Comment by stefanv** [comment]
@acrellin and I were looking at this a bit more; definitely an issue with the NetCDF file being closed prematurely.   I'm checking in with the dask team on the best way to schedule cleanup calls.

File: cesium_app/handlers/model.py [hunk 3]
```diff
@@ -19,6 +19,10 @@
 import xarray as xr
 
 
+def _call_function(func, func_args, func_kwargs,
```
Comment (stefanv) [inline cesium_app/handlers/model.py]: Can we pick a more descriptive name?  Maybe `await_then_execute`?.

How about the following signature:

```
def await_then_execute(func, *func_args, futures=[], **func_kwargs):
```

Then you can call it like:

```
close_future = executor.submit(_await_then_execute, xr.Dataset.close,
                                       fset_data, futures=[computed_model, score_future, save_future])
```

**Comment by acrellin** [comment]
@stefanv I thought of something like that signature, but wondered: if the
futures are list elements rather than standalone arguments, will they will
still be recognized and waited on? But I'll try it that way and see how it
goes.

**Comment by stefanv** [comment]
Urgh, you're probably right.  Let's see, but I'm not optimistic :/

**Comment by acrellin** [comment]
It's green!

**Comment by stefanv** [comment]
@acrellin Suggestion from @mrocklin:

```
@gen.coroutine
def wait_and_call(callback):
    yield _wait([a, b, c])
    callback()

client.loop.add_callback(wait_and_call, myfunc)
```

**Comment by mrocklin** [comment]
```
from distributed.client import _wait
from tornado import gen
```

**Comment by acrellin** [comment]
@mrocklin @stefanv Will give that a try - thanks!

**Comment by stefanv** [comment]
Hrmmm... I didn't expect that to work.  @mrocklin Is this expected, or is this a side-effect we should not rely on?

**Comment by stefanv** [comment]
I guess, to be safe, we could make that a coroutine and await explicitly, like Matt suggested above.

**Comment by mrocklin** [comment]
I don't follow.  What side effect?

**Comment by acrellin** [comment]
@mrocklin I think he's referring to the fact that this (hack) works:
https://github.com/cesium-ml/cesium_web/pull/122/commits/aed9b471551f5f5255922b2d19583e6f89c01a78

**Comment by acrellin** [comment]
more specifically, here https://github.com/cesium-ml/cesium_web/pull/122/commits/aed9b471551f5f5255922b2d19583e6f89c01a78#diff-fcbdd00ee57913aec2f12dc865adb704R22 and here https://github.com/cesium-ml/cesium_web/pull/122/commits/aed9b471551f5f5255922b2d19583e6f89c01a78#diff-fcbdd00ee57913aec2f12dc865adb704R113

**Comment by mrocklin** [comment]
Yeah, that seems reasonable to me.  You're sending along the futures as dummy arguments so that they'll have to exist before the function runs.  I use this sort of trick fairly often.  The only issue here is that you might be causing unnecessary communication.

**Comment by stefanv** [comment]
The unnecessary communication and the implicit side-effect we're using here makes me feel uncomfortable.  Is there any way to wait for a future to complete _without_ shipping its results?

**Comment by mrocklin** [comment]
-  You can do it client side, as is done in the example above using `_wait` and Tornado (probably what I would do) 
-  You can map `lambda *args, **kwargs: None` on the futures and then depend on those.  
-  You could raise an issue in dask/distributed for future work

**Comment by stefanv** [comment]
@acrellin Let's go for Matt's wait and call suggestion, then.  Thanks for the feedback, @mrocklin!

**Comment by acrellin** [comment]
@mrocklin works like a charm, thanks for the suggestion!

**Comment by acrellin** [comment]
@stefanv any further comments?

File: cesium_app/handlers/model.py [hunk 4]
```diff
@@ -102,14 +105,22 @@ def post(self):
...
+            yield _wait(futures_list)
+            return callback(*args)
+
+        model.task_id = save_future.key
         model.save()
 
         loop = tornado.ioloop.IOLoop.current()
-        loop.spawn_callback(self._await_model, future, model)
+        loop.add_callback(_wait_and_call, xr.Dataset.close, fset_data,
+                          futures_list=[computed_model, score_future, save_future])
```
Comment (stefanv) [inline cesium_app/handlers/model.py]: futures_list -> futures

File: cesium_app/models.py [hunk 5]
```diff
@@ -106,6 +108,13 @@ def uris(self):
     def is_owned_by(self, username):
         return self.project.is_owned_by(username)
 
+    def display_info(self):
```
Comment (stefanv) [inline cesium_app/models.py:111]: Docstring.  Also, what does "display info" mean?

**Comment by stefanv** [comment]
Two more minor things, then I think we're good to go.  Feel free to merge when passing.

**Pull request merged**